### PR TITLE
[AutoBuild] Supporting build agents from library; supporting generating agent descriptions

### PR DESCRIPTION
**Pull request by LinxinS97**
## Why are these changes needed?

This PR is to: 

1. Introduces a new feature of AutoBuild: build agents from a library. A library lists domain expert agents' names and system messages. To support this feature, the `build()` function now be separated into three functions: `build()` with no change on API, `build_from_library()`, and `_build_agent()`. The new `build()` function and `build_from_library()` function will generate a list of agent config from scratch or a given library and call the `_build_agent()` to generate agents based on the agent config list. `build_from_library` also supports agent selection from embedding similarity between agent profile and the `building_task`.
2. Supports generating agent descriptions to improve group chat. A description briefly introduces agents that can help the group chat manager judge if this agent should speak in the current group chat round.

I've added a new notebook, example files, and tests for those new features.

## Related issue number

#1092 

## Checks

- [x] I've included any doc changes needed for https://microsoft.github.io/autogen/. See https://microsoft.github.io/autogen/docs/Contribute#documentation to build and test documentation locally.
- [x] I've added tests (if relevant) corresponding to the changes introduced in this PR.
- [x] I've made sure all auto checks have passed.

### CODE REVIEW

**Comment by LinxinS97** [comment]
Hi @sonichi, does the test environment support `gpt-4-1106-preview` or `gpt-4-32k`? We need a gpt with a large context to support the new feature.

File: website/docs/Examples.md [hunk 1]
```diff
@@ -65,6 +65,7 @@ Links to notebook examples:
    - AgentEval: A Multi-Agent System for Assess Utility of LLM-powered Applications - [View Notebook](https://github.com/microsoft/autogen/blob/main/notebook/agenteval_cq_math.ipynb)
 1. **Automatic Agent Building**
    - Automatically Build Multi-agent System with AgentBuilder - [View Notebook](https://github.com/microsoft/autogen/blob/main/notebook/agentchat_autobuild.ipynb)
+   - Automatically Build Multi-agent System with AgentBuilder from library - [View Notebook](https://github.com/microsoft/autogen/blob/main/notebook/agentchat_autobuild_library.ipynb)
```
Comment (JieyuZ2) [inline website/docs/Examples.md:68]: Automatically Build Multi-agent System from Agent Library

File: autogen/agentchat/contrib/agent_builder.py [hunk 2]
```diff
@@ -265,105 +288,253 @@ def clear_all_agents(self, recycle_endpoint: Optional[bool] = True):
     def build(
         self,
         building_task: Optional[str] = None,
-        default_llm_config: Optional[Dict] = None,
         coding: Optional[bool] = None,
-        cached_configs: Optional[Dict] = None,
-        use_oai_assistant: Optional[bool] = False,
         code_execution_config: Optional[Dict] = None,
+        default_llm_config: Optional[Dict] = None,
```
Comment (ekzhu) [inline autogen/agentchat/contrib/agent_builder.py:293]: `default_llm_config` cannot be None because it is later used in `_build` with a `copy()` method.

Comment (ekzhu) [reply autogen/agentchat/contrib/agent_builder.py:293]: Consider either handling the case when `default_llm_config` is `None` or make it a required argument.

**Comment by radman-x** [comment]
Would it make any sense to return a JSON object (possibly by using function call schema) from the AGENT_SYS_MSG_PROMPT llm call to allow it to support defining tool/function calls and, possibly, other future data? I realize this may be a separate PR but may be somewhat related to this one.

**Comment by ekzhu** [comment]
> Would it make any sense to return a JSON object (possibly by using function call schema) from the AGENT_SYS_MSG_PROMPT llm call to allow it to support defining tool/function calls and, possibly, other future data? I realize this may be a separate PR but may be somewhat related to this one.

Does the new function call interface look useful to you? 

https://microsoft.github.io/autogen/docs/Use-Cases/agent_chat/#function-calling

**Comment by radman-x** [comment]
> > Would it make any sense to return a JSON object (possibly by using function call schema) from the AGENT_SYS_MSG_PROMPT llm call to allow it to support defining tool/function calls and, possibly, other future data? I realize this may be a separate PR but may be somewhat related to this one.
> 
> @radman-x Does the new function call interface look useful to you?
> 
> https://microsoft.github.io/autogen/docs/Use-Cases/agent_chat/#function-calling

Yes, that looks good as long as there is a way to use the decorators in AgentBuilder - I don’t see that way since AgentBuilder does not use ConversableAgent nor UserProxyAgent (ie it just uses OpenAIWrapper). Also, of course, there still seem to be some bugs in the decorators that must be fixed. But otherwise, the decorators approach with Python typing and/or Pydantic models is nice.

**Comment by radman-x** [comment]
BTW, on a different subject, it would be really nice if the functions/skills used here as part of AgentBuilder library were compatible with those in AutoGenAssistist/Studio.

**Comment by JieyuZ2** [comment]
hey @radman-x both of your suggestions sound good and maybe you can make an issue for each? we will enhance the autobuild feature PR by PR- and ofc any contribution to autobuild is welcome

**Comment by davorrunje** [comment]
> Yes, that looks good as long as there is a way to use the decorators in AgentBuilder - I don’t see that way since AgentBuilder does not use ConversableAgent nor UserProxyAgent (ie it just uses OpenAIWrapper). Also, of course, there still seem to be some bugs in the decorators that must be fixed. But otherwise, the decorators approach with Python typing and/or Pydantic models is nice.

@radman-x can you please let us know what bugs did you find in the decorators?

**Comment by radman-x** [comment]
> > Yes, that looks good as long as there is a way to use the decorators in AgentBuilder - I don’t see that way since AgentBuilder does not use ConversableAgent nor UserProxyAgent (ie it just uses OpenAIWrapper). Also, of course, there still seem to be some bugs in the decorators that must be fixed. But otherwise, the decorators approach with Python typing and/or Pydantic models is nice.
> 
> @radman-x can you please let us know what bugs did you find in the decorators?

Mainly issue #1061, though that may be fixed in latest repo (haven’t tested it yet). I did have some problems with using register_function() in the agentchat_groupchat_RAG.ipynb example awhile back though I haven’t tested it lately and I realize this is not directly related to using the decorators. So, as long as #1061 is fixed, I do have anything concrete at this time.

I can create issues on these but it will be a week or so since I’m still on holiday (just checking email).

**Review by JieyuZ2 — APPROVED** [review]

**Review by JieyuZ2 — CHANGES_REQUESTED** [review]
need to adapt to #1092

File: test/agentchat/contrib/test_agent_builder.py [hunk 3]
```diff
@@ -2,43 +2,54 @@
...
+        assert agent_config.get("name", None) is not None
+        assert agent_config.get("model", None) is not None
+        assert agent_config.get("description", None) is not None
+        assert agent_config.get("system_message", None) is not None
+
 
 @pytest.mark.skipif(
     skip,
-    reason="openai not installed OR requested to skip",
+    reason="do not run when dependency is not installed",
```
Comment (sonichi) [inline test/agentchat/contrib/test_agent_builder.py:43]:
```suggestion
    reason="do not run when dependency is not installed or requested to skip",
```

File: test/agentchat/contrib/test_agent_builder.py [hunk 4]
```diff
@@ -2,43 +2,54 @@
...
 )
 def test_build():
-    builder = AgentBuilder(config_path=oai_config_path, builder_model="gpt-4", agent_model="gpt-4")
+    builder = AgentBuilder(config_path=OAI_CONFIG_LIST, builder_model="gpt-4", agent_model="gpt-4")
```
Comment (sonichi) [inline test/agentchat/contrib/test_agent_builder.py:46]: "config_path" and "OAI_CONFIG_LIST" don't seem to match semantically. The former is a path and the latter is a list.

File: test/agentchat/contrib/test_agent_builder.py [hunk 5]
```diff
@@ -48,21 +59,78 @@ def test_build():
...
     # check system message
-    for agent, proc in builder.agent_procs_assign.values():
-        assert "TERMINATE" in agent.system_message
+    for cfg in agent_config["agent_configs"]:
+        assert "TERMINATE" in cfg["system_message"]
+
+
+@pytest.mark.skipif(
+    skip or not CHROMADB_INSTALLED,
+    reason="do not run when dependency is not installed",
```
Comment (sonichi) [inline test/agentchat/contrib/test_agent_builder.py:74]:
```suggestion
    reason="do not run when dependency is not installed or requested to skip",
```

File: test/agentchat/contrib/test_agent_builder.py [hunk 6]
```diff
@@ -48,21 +59,78 @@ def test_build():
...
+
+    # check system message
+    for cfg in agent_config["agent_configs"]:
+        assert "TERMINATE" in cfg["system_message"]
 
 
 @pytest.mark.skipif(
     skip,
-    reason="openai not installed OR requested to skip",
+    reason="do not run when dependency is not installed",
```
Comment (sonichi) [inline test/agentchat/contrib/test_agent_builder.py:130]:
```suggestion
    reason="do not run when dependency is not installed or requested to skip",
```

File: test/agentchat/contrib/test_agent_builder.py [hunk 7]
```diff
@@ -86,25 +154,18 @@ def test_save():
...
-    assert saved_configs.get("agent_configs", None) is not None
-    assert saved_configs.get("coding", None) is not None
-    assert saved_configs.get("default_llm_config", None) is not None
+    _config_check(saved_configs)
 
 
 @pytest.mark.skipif(
     skip,
-    reason="openai not installed OR requested to skip",
+    reason="do not run when dependency is not installed",
```
Comment (sonichi) [inline test/agentchat/contrib/test_agent_builder.py:162]: same problem.

File: test/agentchat/contrib/test_agent_builder.py [hunk 8]
```diff
@@ -115,25 +176,17 @@ def test_load():
...
-        assert agent_configs.get(agent_name, None) is not None
-        assert agent_configs[agent_name]["model"] == agent.llm_config["model"]
-        assert agent_configs[agent_name]["system_message"] == agent.system_message
+    _config_check(loaded_agent_configs)
 
 
 @pytest.mark.skipif(
     skip,
-    reason="openai not installed OR requested to skip",
+    reason="do not run when dependency is not installed",
```
Comment (sonichi) [inline test/agentchat/contrib/test_agent_builder.py:186]: same problem

**Review by sonichi — COMMENTED** [review]
Please double check the meaning of skip_openai.

**Comment by LinxinS97** [comment]
Hi @sonichi @ekzhu , I've completed the new features for autobuild. Do you have further comments on that, or can we start the opener tests?

**Review by ekzhu — APPROVED** [review]
Overall I think it is fine for contrib. One comment (not required) is to try to remove unnecessary dependency like chromadb for small-scale vector embedding search -- linear scan is sufficient.

**Comment by LinxinS97** [comment]
> Overall I think it is fine for contrib. One comment (not required) is to try to remove unnecessary dependency like chromadb for small-scale vector embedding search -- linear scan is sufficient.

Thanks! I use chromadb because I see autogen seems to have supported this.

File: test/agentchat/contrib/test_agent_builder.py [hunk 9]
```diff
@@ -2,43 +2,57 @@
 import os
 import json
 import sys
-from packaging.requirements import Requirement
 from autogen.agentchat.contrib.agent_builder import AgentBuilder
-from autogen import UserProxyAgent
 from conftest import skip_openai
 
 sys.path.append(os.path.join(os.path.dirname(__file__), ".."))
```
Comment (sonichi) [inline test/agentchat/contrib/test_agent_builder.py:8]: Should line 10 be before line 8?

**Review by sonichi — COMMENTED** [review]
One minor comment for enabling local test.

**Comment by LinxinS97** [comment]
Hi @gagb @kevin666aa, CompressionTest and GPTAssistantAgent also failed. Could you take a look at that?

**Comment by sonichi** [comment]
The failure for CompressibleAgent is a known issue. @IANTHEREAL for the GPTAssistantAgent failure.

**Comment by sonichi** [comment]
Failure: https://github.com/microsoft/autogen/actions/runs/7429279268/job/20217636681?pr=1039

File: autogen/agentchat/contrib/agent_builder.py [hunk 10]
```diff
@@ -139,26 +192,42 @@ def _create_agent(
...
             system_message: system prompt use to format an agent's behavior.
+            description: a brief description of the agent. This will improve the group chat performance.
             use_oai_assistant: use OpenAI assistant api instead of self-constructed agent.
             world_size: the max size of parallel tensors (in most of the cases, this is identical to the amount of GPUs).
 
         Returns:
             agent: a set-up agent.
         """
-        config_list = autogen.config_list_from_json(self.config_path, filter_dict={"model": [model_name_or_hf_repo]})
+        from huggingface_hub import HfApi
```
Comment (sonichi) [inline autogen/agentchat/contrib/agent_builder.py:205]: The part of using local models is supposed to be moved to the oai layer. @olgavrou for awareness.

**Pull request merged**